### DISCUSSION TITLE
Getting markdown content on collab server

### DISCUSSION THREAD

**Discussion by VincentK-ZA** [Milkdown/milkdown#1394] [Q&A]
When using the collab plugin, it relies internally on the [y-prosemirror](https://github.com/yjs/y-prosemirror) project, meaning the content synced into the yjs doc is an xml fragment.

i.e. if my markdown content is:
```md
# hello world
```
then `yDoc.getXmlFragment("prosemirror").toJSON()` will give:
```html
<heading id="hello-world" level="1">hello world</heading>
```

Is there an easy way to get the markdown content on the server running the yjs connection provider?

I've currently got something working in a very hacky way, whereby I create an instance of the Milkdown editor on the nodejs + express server, relying on [jsdom](https://www.npmjs.com/package/jsdom)

```js
export function setup_jsdom() {
  const dom = new JSDOM(`<!DOCTYPE html><p></p>`);
  global.window = dom.window;
  global.document = dom.window.document;
  global.addEventListener = dom.window.addEventListener;
  global.removeEventListener = dom.window.removeEventListener;
  global.dispatchEvent = dom.window.dispatchEvent;
  global.CustomEvent = dom.window.CustomEvent;
}

export async function create_milkdown_editor() {
  const editor = await Editor.make().use(commonmark).use(gfm).use(emoji).create();

  let serializer = null;
  let schema = null;
  editor.action((ctx) => {
    serializer = ctx.get(serializerCtx);
    schema = ctx.get(schemaCtx);
  });

  return { serializer, schema };
}

export async function get_markdown(doc, serializer, schema) {
  const fragment = doc.getXmlFragment("prosemirror");

  const node = yXmlFragmentToProseMirrorRootNode(fragment, schema);

  const markdown = serializer(node);

  console.log(markdown);
}
```

So I setup jsdom once and create the editor once.

I can then use the `serializer` and `schema` to create the markdown from the `yXmlFragment`

```js
setup_jsdom();
const { serializer, schema } = await create_milkdown_editor();
//...
ysocketio.on("document-update", function (doc, update) {
  get_markdown(doc, serializer, schema);
});
```

---

So this works, but I'm sure there must be a way to setup the necessary plugins and create the serializer and schema only, without needing the view.  

The reason for needing this functionality is since we'd like to only work with the markdown as a source of truth, and have the server periodically persist the markdown to a nosql database.  
I do realize that we would need to ensure that the plugins used on the frontend match those on the backend, else content may get lost.

**Comment by pimsierhuisprorail** [comment]
Ever succeeded in doing this? I want exactly the same, markdown dir as the single source of truth.

**Reply by VincentK-ZA** [reply]
The above mostly worked, though always felt hacky. We ended up moving to using [codemirror](https://codemirror.net/) directly, and also ended up dropping yjs/crdts entirely. Implemented a simple version of what is discussed in [this talk](https://www.youtube.com/watch?v=0d73MdQkJOk) as we did not need offline support and had a central authority server, i.e. we did not need a CRDT.

**Comment by Saul-Mirone** [comment]
I'm also curious about this scenario, I hope someone can give some feedback about how they implement this and how this framework can make some support for this.
